aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +22,333 @@ namespace at {
-        break; \
-    } \
-    ATENSOR##_i = 0; \
+    if (ut.dim == 1)
+      return true;
+
+    /* Reset pointer to beginning of loop */
+    ut.data -= ut.size * ut.stride;
+    for (ut.i = ut.dim - 2; ut.i >= 0; ut.i--) {
+      ut.counter[ut.i]++;
+      /* Jump ahread by the stride of this dimension */
+      ut.data += ut.strides[ut.i];
+
+      if (ut.counter[ut.i] == ut.sizes[ut.i]) {
+        if (ut.i == 0) {
+          return true;
+        } else {
+          /* Reset the pointer to the beginning of the chunk defined by this
+           * dimension */
+          ut.data -= ut.counter[ut.i] * ut.strides[ut.i];
+          ut.counter[ut.i] = 0;
+        }
+      } else
+        return false;
+    }
+    ut.i = 0;
   }
+  return false;
+}
 
 template <typename scalar1, typename scalar2, typename Op>
-void CPU_tensor_apply2_dim(Tensor& tensor1, Tensor& tensor2, int64_t dim, Op op) {
+void CPU_tensor_apply2_dim(
+    Tensor& tensor1,
+    Tensor& tensor2,
+    int64_t dim,
+    Op op) {
   checkBackend("CPU_tensor_apply2", {tensor1, tensor2}, Backend::CPU);
-  bool TH_TENSOR_APPLY_hasFinished = false;
-  int64_t TH_TENSOR_dim_index = 0;
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar1, tensor1, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar2, tensor2, dim, 1)
+  util_tensor<scalar1> ut1(tensor1, dim, true);
+  util_tensor<scalar2> ut2(tensor2, dim, true);
   auto t1_numel = tensor1.numel();
   auto t2_numel = tensor2.numel();
-  if(t1_numel != t2_numel) {
+  if (t1_numel != t2_numel) {
     std::ostringstream oss;
-    oss << "inconsistent tensor size, expected " << tensor1.sizes() << " and " << tensor2.sizes()
-        << " to have the same number of elements, but got " << t1_numel << " and " << t2_numel << " elements respectively";
+    oss << "inconsistent tensor size, expected " << tensor1.sizes() << " and "
+        << tensor2.sizes() << " to have the same number of elements, but got "
+        << t1_numel << " and " << t2_numel << " elements respectively";
     throw std::runtime_error(oss.str());
   }
-  while(!TH_TENSOR_APPLY_hasFinished)
-  {
+  bool TH_TENSOR_APPLY_hasFinished = false;
+  if (tensor1.sizes().equals({0}))
+    TH_TENSOR_APPLY_hasFinished = true;
+  if (tensor2.sizes().equals({0}))
+    TH_TENSOR_APPLY_hasFinished = true;
+  while (!TH_TENSOR_APPLY_hasFinished) {
     /* Loop through the inner most region of the Tensor */
-    for(; tensor1_i < tensor1_size && tensor2_i < tensor2_size; tensor1_i++, tensor2_i++, tensor1_data += tensor1_stride, tensor2_data += tensor2_stride)
-    {
-      op(*tensor1_data, *tensor2_data);
+    for (; ut1.i < ut1.size && ut2.i < ut2.size;
+         ut1.i++, ut2.i++, ut1.data += ut1.stride, ut2.data += ut2.stride) {
+      op(*ut1.data, *ut2.data);
     }
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor1, 0)
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor2, 0)
+    TH_TENSOR_APPLY_hasFinished = tensor_update_counters(ut1);
+    TH_TENSOR_APPLY_hasFinished = tensor_update_counters(ut2);
```

Comment:
This looks weird. Why are you assigning to this value only to overwrite it in the next line?